### PR TITLE
#17717 handle OnComplete

### LINKED ISSUE #17717: Handle OnComplete in ActorSubscriber doc samples
Reported in the [mailing list](https://groups.google.com/forum/#!searchin/akka-user/aren$27t$20shut$20down/akka-user/TeCAVsmpEJI/Qd5HS1wF2aMJ).

http://doc.akka.io/docs/akka-stream-and-http-experimental/1.0-RC3/scala/stream-integrations.html

### PR DESCRIPTION
**Pull request by 2m**
Fixes #17717

### CODE REVIEW

**Comment by drewhk** [comment]
LGTM

**Comment by ktoso** [comment]
The touched tests failed though hm:

```
[info] - illustrate usage of ActorSubscriber *** FAILED *** (3 seconds, 12 milliseconds)
[info]   java.lang.AssertionError: assertion failed: timeout (3 seconds) while expecting 117 messages (got 114)
[info]   at scala.Predef$.assert(Predef.scala:170)
[info]   at akka.testkit.TestKitBase$$anonfun$receiveN_internal$1.apply(TestKit.scala:647)
```

Give it a bit more time IMO, 3 secs may be too tight on jenkins

**Comment by ktoso** [comment]
LGTM otherwise

**Comment by johanandren** [comment]
Still to tight timeouts?

```
[info] - illustrate usage of ActorSubscriber *** FAILED *** (3 seconds, 13 milliseconds)
[info]   java.lang.AssertionError: assertion failed: timeout (3 seconds) while expecting 117 messages (got 116)
[info]   at scala.Predef$.assert(Predef.scala:170)
[info]   at akka.testkit.TestKitBase$$anonfun$receiveN_internal$1.apply(TestKit.scala:647)
[info]   at akka.testkit.TestKitBase$$anonfun$receiveN_internal$1.apply(TestKit.scala:644)
[info]   at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
[info]   at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
[info]   at scala.collection.immutable.Range.foreach(Range.scala:160)
[info]   at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
[info]   at scala.collection.AbstractTraversable.map(Traversable.scala:104)
[info]   at akka.testkit.TestKitBase$class.receiveN_internal(TestKit.scala:644)
[info]   at akka.testkit.TestKitBase$class.receiveN(TestKit.scala:635)
[info]   at akka.testkit.TestKit.receiveN(TestKit.scala:737)
[info]   at docs.stream.ActorSubscriberDocSpec$$anonfun$2.apply(ActorSubscriberDocSpec.scala:90)
[info]   at docs.stream.ActorSubscriberDocSpec$$anonfun$2.apply(ActorSubscriberDocSpec.scala:80)
```

File: akka-docs/rst/scala/code/docs/stream/ActorSubscriberDocSpec.scala [hunk 1]
```diff
@@ -79,11 +82,13 @@ class ActorSubscriberDocSpec extends AkkaSpec {
 
     //#actor-subscriber-usage
     val N = 117
-    Source(1 to N).map(WorkerPool.Msg(_, replyTo))
+    val worker = Source(1 to N).map(WorkerPool.Msg(_, replyTo))
       .runWith(Sink.actorSubscriber(WorkerPool.props))
     //#actor-subscriber-usage
 
+    watch(worker)
     receiveN(N).toSet should be((1 to N).map(WorkerPool.Done).toSet)
```
Comment (patriknw) [inline akka-docs/rst/scala/code/docs/stream/ActorSubscriberDocSpec.scala]: The new test failure is not about too short timeout. In the log you can see
```
[WARN] [08/24/2016 12:48:21.862] [ActorSubscriberDocSpec-akka.actor.default-dispatcher-2] [akka://ActorSubscriberDocSpec/user/StreamSupervisor-46/flow-0-1-actorSubscriberSink] received dead letter from Actor[akka://ActorSubscriberDocSpec/user/StreamSupervisor-46/flow-0-1-actorSubscriberSink/$b#1432547271]: Reply(116)
[WARN] [08/24/2016 12:48:21.863] [ActorSubscriberDocSpec-akka.actor.default-dispatcher-2] [akka://ActorSubscriberDocSpec/user/StreamSupervisor-46/flow-0-1-actorSubscriberSink] received dead letter from Actor[akka://ActorSubscriberDocSpec/user/StreamSupervisor-46/flow-0-1-actorSubscriberSink/$c#2143819031]: Reply(117)
```

i.e. the actor was stopped before the replies were handled.

In `case OnComplete` you can only stop if queue is empty, and in `case Reply` you should stop `if (canceled && queue.isEmpty`

Comment (patriknw) [inline akka-docs/rst/scala/code/docs/stream/ActorSubscriberDocSpec.scala]: @2m what's the status here? did you see my comment?

Comment (2m) [inline akka-docs/rst/scala/code/docs/stream/ActorSubscriberDocSpec.scala]: Thanks for reminding me. I have seen it and remember making changes. Changes are unpushed for some reason. I will get back to this right now.

File: akka-docs/rst/java/code/docs/stream/ActorSubscriberDocTest.java [hunk 2]
```diff
@@ -170,18 +172,26 @@ public WorkerPool() {
 
             if (queue.size() > MAX_QUEUE_SIZE)
               throw new RuntimeException("queued too many: " + queue.size());
-                  
+
             router.route(WorkerPoolProtocol.work(msg.id), self());
         }).
+        match(ActorSubscriberMessage.onCompleteInstance().getClass(), complete -> {
```
Comment (johanandren) [inline akka-docs/rst/java/code/docs/stream/ActorSubscriberDocTest.java]: Wouldn't `matchEquals(ActorSubscriberMessage.onCompleteInstance(), ...` be nicer (and more aligned with what you'd do in Scala) here?

**Comment by johanandren** [comment]
LGTM

**Pull request merged**